api/documents/index.ts

```diff
@@ -0,0 +1,47 @@
+// Time configuration
+const INITIAL_TIMEOUT_PERIOD = 5000; // 5 seconds
+const MAX_TIMEOUT_PERIOD = 180000; // 180 seconds
+
+let isRequestRateLimited = false;
+
+// Handle requests being sent to the target server
+export function handleRequest(req, res, proxyRes) {
+  const defaultTimeoutPeriod = INITIAL_TIMEOUT_PERIOD;
+
+  // Try to retrieve the next timeout period from the session
+  const nextTimeout = res.session.nextTimeout;
+
+  // If there is no timeout period in the session, set it to the default
+  const timeoutPeriod = nextTimeout || defaultTimeoutPeriod;
+
+  // Check if the last document upload was completed within the timeout period
+  const lastUploadTime = res.session.lastUploadTime || 0;
+  const elapsedTime = Date.now() - lastUploadTime;
+
+  if (elapsedTime < timeoutPeriod) {
+    isRequestRateLimited = true;
+    proxyRes.status(429).send({ message: 'Too many requests' });
+
+    return false;
+  }
+
+  return true;
+}
+
+// Handle responses being sent back to the client
+export function handleResponse(req, res, proxyRes, json) {
+  // Current time stored as the last time a document upload was completed
+  res.session.lastUploadTime = Date.now();
+
+  // Double the timeout period up to the maximum, if not rate-limited
+  if (!isRequestRateLimited) {
```

Comment:
If we get another request while we are within the rate limit timeout, then double the timeout, up to some max value. The isRequestRateLimited value (which needs to be session specific, not global) is set to true if we get a request within the timeout period. Therefore I think this needs to be inverted - it should say:
 if (req.session.isRequestRateLimited)